### PR TITLE
Redirect_uri - Prohibit the use of fragment and better handle query component

### PR DESCRIPTION
**Pull request by PGBI**
The oauth2 specification states that a client redirect_uri may include a query component, and must not contain a fragment component (see: https://tools.ietf.org/html/rfc6749#section-3.1.2)

Currently Kong does not comply with this rule, and generates a non valid redirect_uri if the client redirect_uri contains a query component or a fragment.

### Fragment handling:
If a client redirect_uri is of the form http://domain.tld/path#fragment, `oauth2/authorize` endpoint will currently return:
```json
{ "redirect_uri" : "http://domain.tld/path#fragment?code=xxx" }
```
With this PR, it would return:
```json
{ 
"error" : "invalid_request", 
"error_description": "Fragment not allowed in redirect_uri" 
}
```

### Query component handling
If a client redirect_uri is of the form http://domain.tld/path?foo=bar, `oauth2/authorize` endpoint will currently return:
```json
{ "redirect_uri" : "http://domain.tld/path?foo=bar?code=xxx" }
```
With this PR, it would return:
```json
{ "redirect_uri" : "http://domain.tld/path?foo=bar&code=xxx" }
```

### CODE REVIEW

File: kong/plugins/oauth2/access.lua [hunk 1]
```diff
@@ -140,8 +139,15 @@ local function authorize(conf)
 
       -- Check client_id and redirect_uri
       redirect_uri, client = get_redirect_uri(parameters[CLIENT_ID])
+      parsed_redirect_uri = require("socket.url").parse(redirect_uri)
```
Comment (thibaultCha) [inline kong/plugins/oauth2/access.lua]: Better to require `socket.url` at the module level instead of here.

File: kong/plugins/oauth2/access.lua [hunk 2]
```diff
@@ -172,9 +178,17 @@ local function authorize(conf)
   -- Adding the state if it exists. If the state == nil then it won't be added
   response_params.state = state
 
+  -- Appending kong generated params to redirect_uri query string
+  if parsed_redirect_uri then
+    if not parsed_redirect_uri.query then
+      parsed_redirect_uri.query = ""
+    end
+    parsed_redirect_uri.query = ngx.encode_args(utils.table_merge(ngx.decode_args(parsed_redirect_uri.query), response_params))
```
Comment (thibaultCha) [inline kong/plugins/oauth2/access.lua]: Maybe better to use `utils.encode_args` as it preserves percent-encoding of the querystring. See #749

**Comment by thibaultCha** [comment]
That's really nice, thank you! Just two small modifications to be done and it should be good. Also I'm realizing this plugin and its test suite **absolutely** need a cleanup!

**Comment by thibaultCha** [comment]
PS: if you rebase on Mashape:next, you should be able to have proper CI and ensure that your tests pass on Travis.

**Comment by PGBI** [comment]
Thanks @thibaultCha for your feedback. I updated the PR accordingly.

**Comment by thibaultCha** [comment]
Looks good to me, I reopened #930 with a squashed version of your patch. Thanks!

**Pull request closed without merging**